### ISSUE TITLE
`z.input<…>` of coercibles should be `any`

### ISSUE THREAD

**Issue by twiddler** [colinhacks/zod#2519]
In this code …

```ts
import { z } from 'zod'

const b = z.coerce.boolean()

type B = z.input<typeof b>
```

… `B` evaluates to …

```ts
type B = boolean;
```

I expected `B` to evaluate to …

```ts
type B = any;
```

If this is intended behaviour, I'd be happy if you could share your reasoning. Thank you. :pray: 

This is related to #2421.

zod version: v3.21.4

**Comment by iway1** [comment]
I think `z.input` means "The type of a valid input" in the context of zod (the type before any transforms etc are made).

I believe you're correct that the input should include more types than just the type that the value is being coerced to.

But not all coercibles accept any type, for example if you do

```ts
// Errors
const parsed = z.coerce.number().parse("no");

// parsed is now the number 0
const parsed = z.coerce.number().parse(null);
```

So for `z.coerce.number()`, it seems that `z.input<...>` should be `number | string | null` since those are each input types that could parse successfully (not sure if that's all the types).

For `.string()` and `.boolean()` I think that `any` type should be returned from `z.input` since in JS everything can be converted to a string or bool

**Comment by akomm** [comment]
I don't say it is bad do it based on spec, but I think first-place it should be based on what coerce of this library can work with. So if following the spec, then we need to make sure that coerce works based on spec, then the z.input would end up being exactly in line with it.

**Comment by redbmk** [comment]
It looks like this also makes it so that `z.coerce.boolean().default(false)` doesn't work as expected. And using `null`, `undefined`, or `0` instead gives a typescript error.

This is what I'm doing as a workaround:

```typescript
z.object({
  SOME_BOOL_VALUE: z.string().default("false").transform(value => value !== "false"),
}).parse(process.env);
```

As a side note, since this is coming from an env, even something like `SOME_BOOL_VALUE=0` or `false` would really be a string and get treated as `true` using `coerce.boolean()`, so maybe this is the best we can do for booleans coming from strings.

**Comment by twiddler** [comment]
@redbmk Unrelated to the original issue: Since you default to `"false"`, it seems like that might be the "less risky" setting for your application, so maybe consider `.transform(value => value === "true")` such that your boolean is only ever true when one really intends to.

**Comment by DaniFoldi** [comment]
I ran into this recently, ended up using a union of primitive types and calling the appropriate constructor in `.transform` manually.

Posting it here as a workaround until this gets fixed in `z.coerce`:
```ts
import { z } from 'zod'

const zStringToNumber = z.coerce.number()

type zIn1 = z.input<typeof zStringToNumber>
//   ^?

const zStringToNumber2 = z.union([
  z.string(), // maybe NaN
  z.number(),
  z.boolean(), // 0 or 1
  z.bigint(),
  z.date(),
  z.symbol(), // throws
  z.undefined(), // NaN
  z.null() // 0
]).transform(v => Number(v))

type zIn2 = z.input<typeof zStringToNumber2>
//   ^?
```
[TS playground](https://www.typescriptlang.org/play?target=99&moduleResolution=99#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgFBkDGEAdgM7zIDKMUwNA5gCoQByAriABGAUyhwAvCgB01MZRHSag0VAAUASgowAnmBEoAkjQCMkmezD8YAHl36ImFCzaceA4WIB8ZAPS+4QIA9AH4KanpGF3ZuPhUxACZzZGl+GmBaNQBtMkCUhlcOTQAaOH84EABDHVE4XkreXJllT3UNYqaUoQgIABsRSpoSsoCABjhoOBNO6SFgDnYYEpniSpgRZbzpOh1hPuHymAALHAB3Ohm04hFMdhFiA4D6xq3lXt7NEbhRsgBdDWkrEGdEw0BAagAbpIvHV4uoIRotGR7AZkMYklIUpZrHY9CJHM5WDF3HCEj5ysEwkA)

**Comment by colinhacks** [comment]
This is true. It was a tradeoff I made so I could ship `z.coerce` without a breaking change. Because `ZodString` has no generics, there's no way for me to change its `_input` type without adding one, which would have been breaking. Another option would be to introduce some kind of `ZodCoerce` type, but then you wouldn't be able to method calls `z.coerce.string().min(5)`, etc.

This will be fixed in Zod 4 but you can work around it with `.transform()` and `.pipe()` as others have suggested.

**Comment by colinhacks** [comment]
The input type of all "coercibles" is now `unknown` in Zod 4.
